src/Typedefs/Idris.idr

```diff
@@ -0,0 +1,260 @@
+  showMu tvars td (Inn x) =
+    let printMu = (Mu' [] (map DPair.fst tvars) td ** assert_total $ showMu tvars td)
+     in parens' (assert_total $ showTy (printMu::tvars) td x)
+
+  ||| Print the content of a TDef provided functions to display Types as strings
+  ||| @tvars a vector of n functions that map types to their string representation
+  ||| @td the tdef to show
+  ||| @x the idris type represented by `td` using `tvars` for free variables
+  showTy : (tvars : Vect n (a : Type ** a -> String))
+        -> (td : TDefR n)
+        -> (x : Ty' [] (map DPair.fst tvars) td)
+        -> String
+  showTy                  tvars  T0                          x        impossible
+  showTy                  tvars  T1                          x        = show x
+  showTy                  tvars  (TSum [a,b])               (Left x)  = "Left " ++ parens' (showTy tvars a x)
+  showTy                  tvars  (TSum [a,b])               (Right x) = "Right " ++ parens' (showTy tvars b x)
+  showTy                  tvars  (TSum (a::b::c::xs))       (Left x)  = "Left " ++ parens' (showTy tvars a x)
+  showTy                  tvars  (TSum (a::b::c::xs))       (Right x) = "Right " ++ parens' (assert_total $ showTy tvars (TSum (b::c::xs)) x)
+  showTy   ((a ** showA)::tvars) (TVar FZ)                   x        = showA x
+  showTy {n=S (S n')} (_::tvars) (TVar (FS i))               x        = showTy {n = S n'} tvars (TVar i) x
+  showTy                  tvars  (TMu m)                     x        = showMu tvars (args m) x
+  showTy   ((a ** showA)::tvars) (RRef FZ)                   x        = showA x
+  showTy {n=S (S n')} (_::tvars) (RRef (FS i))               x        = showTy {n = S n'} tvars (RRef i) x
+  showTy                  tvars  (TApp (TName name def) xs)  x
+    = assert_total $ showTy tvars (def `ap` xs) (convertTy x)
+  showTy {n}              tvars  (TProd xs)                  x
+    = parens $ concat $ List.intersperse ", " (showProd xs x)
+    where showProd : (ys : Vect (2 + k) (TDefR n))
+                  -> Tnary' [] (map DPair.fst tvars) ys Pair -> List String
+          showProd [a, b]        (x, y) = (showTy tvars a x)::[showTy tvars b y]
+          showProd (a::b::c::ys) (x, y) = (showTy tvars a x)::showProd (b::c::ys) y
+
+-- Can't replace sp by [] for some reason
+Show (Mu' sp [] td) where
+  show y {sp=[]} = showMu [] td y
+  show y {sp=sp} = "cannot show when specialisation context is non-empty"
+
+-------------------------------------------------------------------------------
+-- Type Equality                                                             --
+-------------------------------------------------------------------------------
+-- Note: Equality on Types only work when the specialisation context is empty--
+-------------------------------------------------------------------------------
+
+mutual
+
+  eqMu : (tvars : Vect n (a : Type ** a -> a -> Bool)) -> (td : TDefR (S n)) ->
```

Comment:
you could do all these for non-`[]` specialisations if you ask for the corresponding data for the specialisation... but for now maybe its enough to note this in a comment

Reply:
Actually I tried and there was a problem because of the substitution mechanism in `TApp` with `ap`. Maybe we don't need to do the substitution? I'll put a comment about that.